### PR TITLE
[DPE-4179, DPE-4219] Add integration test for upgrades

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We do not have integration tests for upgrade

## Solution
Add integration tests

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]
Wow, very nice! Thank you!

P.S. I would add in the future: deploy stable, upgrade to edge(PR). But sure... after stable VM release.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -158,13 +158,18 @@ jobs:
...
+      #       {
+      #         "UBUNTU_PRO_TOKEN" : "${{ secrets.UBUNTU_PRO_TOKEN }}",
+      #         "LANDSCAPE_ACCOUNT_NAME": "${{ secrets.LANDSCAPE_ACCOUNT_NAME }}",
+      #         "LANDSCAPE_REGISTRATION_KEY": "${{ secrets.LANDSCAPE_REGISTRATION_KEY }}",
+      #       }
+      - name: Setup tmate session
+        uses: mxschmitt/action-tmate@v3
+        with:
+          limit-access-to-actor: true
+          detached: true
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: the commit was intentional -- part of a debug session. removed in [74e35ae](https://github.com/canonical/mysql-router-operator/pull/135/commits/74e35ae9383c27ab584d104366dbbe5974436ff1)

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -54,3 +61,22 @@ def juju_has_secrets(mocker: MockerFixture, request):
             JujuVersion, "has_secrets", new_callable=PropertyMock
         ).return_value = True
         return True
+
+
+@pytest.fixture()
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/conftest.py]: addressed in [218555c](https://github.com/canonical/mysql-router-operator/pull/135/commits/218555c7aa5ce95ddfa7edd320ebde523109e845)

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -38,10 +45,7 @@ async def get_inserted_data_by_application(unit: Unit) -> str:
     Returns:
         A string representing the inserted data
     """
-    action = await unit.run_action("get-inserted-data")
-    result = await action.wait()
-
-    return result.results.get("data")
+    return (await run_action(unit, "get-inserted-data")).get("data")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/helpers.py:50]: addressed in [218555c](https://github.com/canonical/mysql-router-operator/pull/135/commits/218555c7aa5ce95ddfa7edd320ebde523109e845)

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -242,3 +248,175 @@ async def get_tls_certificate_issuer(
...
+    This enables us to retrieve the name of the deployed application in an existing model.
+
+    Note: if multiple applications with the application name exist,
+    the first one found will be returned.
+    """
+    for application in ops_test.model.applications:
+        if application_name_substring == application:
+            return application
+
+    return ""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/helpers.py:267]: it is a helper function that is imported from mysql vm. i'd like to keep as many helpers the same as we can, so we can easily refactor them into common code in the future if necessary

File: tests/integration/test_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        workload_version_file.write(f"{old_workload_version}+testupgrade\n")
+        workload_version_file.flush()
+
+    yield
+
+    logger.info("Restoring workload_version")
+
+    with open("workload_version", "w") as workload_version_file:
+        workload_version_file.write(f"{old_workload_version}\n")
+        workload_version_file.flush()
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: removed fixture in [218555c](https://github.com/canonical/mysql-router-operator/pull/135/commits/218555c7aa5ce95ddfa7edd320ebde523109e845). but used `pathlib.Path` and `read_text` where applicable

File: tests/integration/test_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        stop=tenacity.stop_after_delay(RETRY_TIMEOUT),
+        wait=tenacity.wait_fixed(10),
+    ):
+        with attempt:
+            workload_version = await get_workload_version(ops_test, mysql_router_unit.name)
+            assert old_workload_version != workload_version
+
+    await ops_test.model.wait_for_idle(
+        apps=[MYSQL_APP_NAME], status="active", idle_period=30, timeout=TIMEOUT
+    )
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: great catch! changed in [218555c](https://github.com/canonical/mysql-router-operator/pull/135/commits/218555c7aa5ce95ddfa7edd320ebde523109e845)

File: tests/integration/test_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py:132]: refactored the test to deploy 3 router units so we can test for `resume-upgrade`

File: tests/integration/test_upgrade.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    logger.info("Ensure continuous writes after rollback procedue")
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+    os.remove(fault_charm)
+
+
+def inject_invalid_workload_version(charm_file: typing.Union[str, pathlib.Path]) -> None:
+    """Inject an invalid charm_version file into the mysqlrouter charm."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: added comments, plus added docstrings where i thought there might be confusion about the flow of the tests

File: tests/integration/test_upgrade.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+    await ops_test.model.relate(
+        f"{MYSQL_ROUTER_APP_NAME}:backend-database", f"{MYSQL_APP_NAME}:database"
+    )
+    await ops_test.model.relate(f"{TEST_APP_NAME}:database", f"{MYSQL_ROUTER_APP_NAME}:database")
+
+    logger.info("Waiting for applications to become active")
+    # We can safely wait only for test application to be ready, given that it will
+    # only become active once all the other applications are ready.
+    await ops_test.model.wait_for_idle([TEST_APP_NAME], status="active", timeout=TIMEOUT)
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tests/integration/test_upgrade.py]: checking all apps active in af7cce6ea4fc7dee23dcf705b871ea3b5d2bcd0d

File: tests/integration/test_upgrade.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply tests/integration/test_upgrade.py:134]: ah, that's why

**Review by paulomach — APPROVED** [review]
Nice

**Pull request merged**